src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/WpfComWrappers.TfThreadMgrWrapper.cs

```diff
@@ -0,0 +1,508 @@
+        {
+            fixed (MSG* msgPtr = &msg)
+            fixed (int* resultPtr = &result)
+            {
+                var hr = ((delegate* unmanaged<IntPtr, MSG*, IntPtr, int, int, int*, int>)(*(*(void***)_messagePumpInstance + 4)))
+                    (_messagePumpInstance, msgPtr, hwnd, msgFilterMin, msgFilterMax, resultPtr);
+                if (NativeMethods.Failed(hr))
+                {
+                    Marshal.ThrowExceptionForHR(hr);
+                }
+            }
+        }
+
+        public void PeekMessageW(ref MSG msg, IntPtr hwnd, int msgFilterMin, int msgFilterMax, int removeMsg, out int result)
+        {
+            fixed (MSG* msgPtr = &msg)
+            fixed (int* resultPtr = &result)
+            {
+                var hr = ((delegate* unmanaged<IntPtr, MSG*, IntPtr, int, int, int, int*, int>)(*(*(void***)_messagePumpInstance + 5)))
+                    (_messagePumpInstance, msgPtr, hwnd, msgFilterMin, msgFilterMax, removeMsg, resultPtr);
+                if (NativeMethods.Failed(hr))
+                {
+                    Marshal.ThrowExceptionForHR(hr);
+                }
+            }
+        }
+
+        public void GetMessageW(ref MSG msg, IntPtr hwnd, int msgFilterMin, int msgFilterMax, out int result)
+        {
+            fixed (MSG* msgPtr = &msg)
+            fixed (int* resultPtr = &result)
+            {
+                var hr = ((delegate* unmanaged<IntPtr, MSG*, IntPtr, int, int, int*, int>)(*(*(void***)_messagePumpInstance + 6)))
+                    (_messagePumpInstance, msgPtr, hwnd, msgFilterMin, msgFilterMax, resultPtr);
+                if (NativeMethods.Failed(hr))
+                {
+                    Marshal.ThrowExceptionForHR(hr);
+                }
+            }
+        }
+
+        public void AdviseSink(ref Guid riid, [MarshalAs(UnmanagedType.Interface)] object obj, out int cookie)
+        {
+            IntPtr unknownPtr = Marshal.GetIUnknownForObject(obj);
+            fixed (Guid* iidPtr = &riid)
+            fixed (int* cookiePtr = &cookie)
+            {
+                var hr = ((delegate* unmanaged<IntPtr, Guid*, IntPtr, int*, int>)(*(*(void***)_sourceInstance + 3)))
+                    (_sourceInstance, iidPtr, unknownPtr, cookiePtr);
+                if (NativeMethods.Failed(hr))
+                {
+                    Marshal.ThrowExceptionForHR(hr);
+                }
+            }
+        }
+
+        public void UnadviseSink(int cookie)
+        {
+            var hr = ((delegate* unmanaged<IntPtr, int, int>)(*(*(void***)_sourceInstance + 4)))
+                (_sourceInstance, cookie);
+            if (NativeMethods.Failed(hr))
+            {
+                Marshal.ThrowExceptionForHR(hr);
+            }
+        }
+
+        public void AdviseKeyEventSink(int clientId, [MarshalAs(UnmanagedType.Interface)] object obj, [MarshalAs(UnmanagedType.Bool)] bool fForeground)
+        {
+            IntPtr unknownPtr = Marshal.GetIUnknownForObject(obj);
```

Comment:
Leaking QI.

Reply:
How it is leaking here? Are you mean in case of any failures inside unmanaged part, we should release `unknownPtr` ?

Reply:
Or do you mean, once we pass COM pointer to unamanged side we can safely release all instances from unmanaged part?

Reply:
The `Marshal.GetIUnknownForObject` and `Marshal.QueryInterface` APIs both should perform an `AddRef`. If the API being called with the `IUnknown` instances needs to retain ownership it should call its own `AddRef`. Which means unless it can be confirmed that the calling API takes over ownership of the `IUnknown` and will call `Release` it is a leak.

Reply:
Then I would say I take care of issues. At least I think so

Reply:
Should this be this?
```cs
IntPtr unknownPtr = default;
try
{
    unknownPtr = Marshal.GetIUnknownForObject(obj);
    ....
```